### DISCUSSION TITLE
Wrongly created/initialized async session leads to hard crash ?

### DISCUSSION THREAD

**Discussion by jpc1976** [sqlalchemy/sqlalchemy#10553] [Usage Questions] [answered]
I am trying to get SQLAlchemy async sessions to work (ultimately within a FastAPI app) but the interpreter keeps crashing as soon as an sql error is encountered. I managed to get the same behaviour with a minimalistic example:

```py
import asyncio

import sqlalchemy
from sqlalchemy import insert, NullPool
from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession, async_sessionmaker
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column

class Base(DeclarativeBase):
    pass

# create table test_table
# (
#     a varchar(4) not null,
#     b integer not null,
#
#     primary key(a, b)
# );

class TestTable(Base):
    __tablename__ = "test_table"

    a: Mapped[str] = mapped_column(primary_key=True)
    b: Mapped[int] = mapped_column(primary_key=True)

def _build_db_url():
    return sqlalchemy.URL.create(
        "postgresql+asyncpg",
        username='scott',
        password='tiger',
        host='localhost',
        database='test')

engine = create_async_engine(_build_db_url(), future=True, echo=False)

async def do_sql(a: str, b: int):
    stmt = insert(TestTable).values(a=a, b=b)
    session = async_sessionmaker(
        bind=engine,
        class_=AsyncSession,
        autoflush=True,
        expire_on_commit=False)
    async with session() as s:
        try:
            await s.execute(stmt)
        except Exception as e:
            print(f"Caught {e}")
        finally:
            await s.commit()
            await s.close()
            await engine.dispose()

if __name__ == "__main__":
    asyncio.run(do_sql('lala', 3))
    asyncio.run(do_sql('lalala', 5))
```

The first call works and inserts the row, but the second one (intentionally) doesn't work since the string 'lalala' is longer than the max length of the column (i.e. varchar(4)), which means that PG generates an error. In the second case, I'd expect to catch an SQL-related exception, but instead I get a hard crash:

```
Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)
```

I also tried using psycopg instead of asyncpg as the driver, but I get the same behaviour.

Am I doing something wrong in the way I create and use the async sessions ? Any feedback appreciated!

I am using:

sqlachemy 2.0.22 
asyncpg 0.28.0
psycopg 3.1.12
postresql 15.1

on MacOS 10.15.7 (x86_64)

**Comment by CaselIT** [comment]
Hi,

Things are working fine on this side on both windows and linux.
Note that calling multiple `asyncio.run` may lead to issues, since each one will create a new loop (IIRC) and in general you can't share async things between loops.

Try doing the following
```py
async def go():
    await (do_sql('lala', 3))
    await (do_sql('lalala', 5))
if __name__ == "__main__":
    asyncio.run(go())
```

also a better execution block would be the following:
```py
    try:
        async with session() as s:
            await s.execute(stmt)
            await s.commit()
    except Exception as e:
        print(f"Caught {e}")
    finally:
        await engine.dispose()
```
- you commit of a aborted transaction will lead to another error
- there is no need of calling close since the context manager does that for you
- dispose runs also if the commit raises an error and it's after the context manager has closed the session

Overall sqlalchemy, asyncpg and greenlep all are compiled extensions. Maybe try checking if recreating the env will fix this issue.

What python version are you using?

**Comment by jpc1976** [comment]
Thanks for the feedback, I'll be looking a native extensions (perhaps reinstalling/recompiling them).

I am using Python 3.12.0

**Reply by CaselIT** [reply]
I forgot to mention that I used python 3.12 on windows and 3.11 on linux in the message above

**Reply by jpc1976** [reply]
Still the same issue on MacOS after reinstalling the environment. I also upgraded to PG 16.0, but it doesn't solve the problem.

There seems to be something in the stack I am using which doesn't like handling postgres errors in async mode...

**Reply by CaselIT** [reply]
that's strange. Can you create a new env, with just sqlalchemy asyncpg / psycopg (and the dependencies they install of course) and run just the script you posted in the initial discussion?

maybe try using python 3.11 too to check if something changes

**Reply by jpc1976** [reply]
I created a separate environment from scratch using only those dependencies:

```
> pip list
Package           Version
----------------- -------
asyncpg           0.28.0
greenlet          3.0.1
pip               23.3.1
psycopg           3.1.12
SQLAlchemy        2.0.22
typing_extensions 4.8.0
```

I ran the above script again in this environment, still getting a hard crash:

```
Using Python 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 12:29:27) [Clang 14.0.6 ]
Segmentation fault: 11
```

I also tried to rollback to python 3.11, but I get the same behaviour:

```
Using Python 3.11.5 (main, Sep 11 2023, 08:19:27) [Clang 14.0.6 ]
Segmentation fault: 11
```

FWIW, I also ran the script in valgrind, here's the error which triggers the segfault:

```
==48137== Invalid read of size 8
==48137==    at 0x10126B936: __cxxabiv1::scan_eh_tab(__cxxabiv1::(anonymous namespace)::scan_results&, _Unwind_Action, bool, _Unwind_Exception*, _Unwind_Context*) (in /usr/lib/libc++abi.dylib)
==48137==    by 0x10126B584: __gxx_personality_v0 (in /usr/lib/libc++abi.dylib)
==48137==    by 0x101872AA0: _Unwind_RaiseException (in /usr/lib/system/libunwind.dylib)
==48137==    by 0x10126B160: __cxa_throw (in /usr/lib/libc++abi.dylib)
==48137==    by 0x103CA5A14: greenlet::Greenlet::g_switch_finish(greenlet::Greenlet::switchstack_result_t const&) (in /Users/scott/temp/sqlalchemy/crash/.venv/lib/python3.11/site-packages/greenlet/_gree
nlet.cpython-311-darwin.so)
==48137==    by 0x103CA236E: greenlet::MainGreenlet::g_switch() (in /Users/scott/temp/sqlalchemy/crash/.venv/lib/python3.11/site-packages/greenlet/_greenlet.cpython-311-darwin.so)
==48137==    by 0x103CA350C: green_switch(_greenlet*, _object*, _object*) (in /Users/scott/temp/sqlalchemy/crash/.venv/lib/python3.11/site-packages/greenlet/_greenlet.cpython-311-darwin.so)
==48137==    by 0x100085E43: method_vectorcall_VARARGS_KEYWORDS (in /Users/scott/temp/sqlalchemy/crash/.venv/bin/python)
==48137==    by 0x1001CCEF6: _PyEval_EvalFrameDefault (in /Users/scott/temp/sqlalchemy/crash/.venv/bin/python)
==48137==    by 0x10007389E: _PyFunction_Vectorcall (in /Users/scott/temp/sqlalchemy/crash/.venv/bin/python)
==48137==    by 0x10007975B: method_vectorcall (in /Users/scott/temp/sqlalchemy/crash/.venv/bin/python)
==48137==    by 0x100072F93: PyObject_Call (in /Users/scott/temp/sqlalchemy/crash/.venv/bin/python)
==48137==  Address 0x1000007feedfaef is not stack'd, malloc'd or (recently) free'd
```

Perhaps someone will be able to interpret it.

**Reply by CaselIT** [reply]
I'm no expert at all, but this seems a greenlet crash. Can you try in 3.11 to install `greenlet<3`?

Thanks for testing this

**Reply by jpc1976** [reply, marked as answer]
I downgraded the greenlet version (I used the latest release before  < v3):

```
pip install --force-reinstall greenlet==2.0.2
```

And this time it works, I get the SQL exception I was expecting without it crashing!

```
Using Python 3.11.5 (main, Sep 11 2023, 08:19:27) [Clang 14.0.6 ]
Caught (sqlalchemy.dialects.postgresql.asyncpg.Error) <class 'asyncpg.exceptions.StringDataRightTruncationError'>: value too long for type character varying(4)
[SQL: INSERT INTO test_table (a, b) VALUES ($1::VARCHAR, $2::INTEGER)]
[parameters: ('lalala', 5)]
(Background on this error at: https://sqlalche.me/e/20/dbapi)
```

Is there a downside in (at least temporarily) using greenlet 2.0.2 with SQLAlchemy ?

**Reply by CaselIT** [reply]
None when using using python < 3.12.
For python 3.12 I don't think it's compatible with greenlet version before 3.x, so you would be limited there

**Comment by zzzeek** [comment]
Hi -

we need to consider greenlet here, which is the other big native dependency.   How are you installing greenlet and what's the full python 3.12 version?

**Reply by jpc1976** [reply]
Hi,

Here's the output of the sys.version command:

```
Using Python 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 12:29:27) [Clang 14.0.6 ]
```

Note that, being on a old MacBook that's still running Catalina, I have an older version of clang installed (version 12 being the last one officially supported by Apple on Catalina):

```
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Could there be some binary incompatibilities w.r.t. default libraries ?

Also, here are all the versions of the packages that have been installed in this environment (obtained using 'poetry show')

```
annotated-types     0.6.0     
anyio               3.7.1     
asgi-correlation-id 4.2.0     
asttokens           2.4.1     
asyncpg             0.28.0    
attrs               23.1.0    
cattrs              23.1.2    
certifi             2023.7.22 
click               8.1.7     
colorama            0.4.6     
elementpath         4.1.5     
executing           2.0.1     
fastapi             0.104.1   
greenlet            3.0.1     
h11                 0.14.0    
httpcore            0.18.0    
httpx               0.25.0    
icecream            2.1.3     
idna                3.4       
iniconfig           2.0.0     
jproperties         2.1.1     
jsonpickle          3.0.2     
orjson              3.9.10    
packaging           23.2      
pluggy              1.3.0     
psycopg             3.1.12    
psycopg-binary      3.1.12    
psycopg-pool        3.1.8     
psycopg2            2.9.9     
pydantic            2.4.2     
pydantic-core       2.10.1
pygments            2.16.1    
pytest              7.4.3     
pytest-asyncio      0.21.1    
python-dotenv       1.0.0     
setuptools          68.2.2    
six                 1.16.0    
sniffio             1.3.0     
sqlalchemy          2.0.22    
starlette           0.27.0    
structlog           23.2.0    
typing-extensions   4.8.0     
uvicorn             0.23.2    
xmlschema           2.5.0     
xmltodict           0.13.0    
```

Greenlet must have been installed as a co-dependency (perhaps of SQLAlchemy?), as I did not install it directly. It might be the source of the problem I am experimenting, as some of the crashes seem to occur in this snippet (in session.py, line 454 and following)
```
result = await greenlet_spawn(
            self.sync_session.execute,
            statement,
            params=params,
            execution_options=execution_options,
            bind_arguments=bind_arguments,
            **kw,
        )
```

I must admit that I am not sure of what greenlet exactly does in that case.

**Reply by zzzeek** [reply]
try running a simple script from their docs at https://greenlet.readthedocs.io/en/latest/greenlet.html#the-greenlet-lifecycle


```py
>>> from greenlet import greenlet

>>> def test1():
...     print("[gr1] main  -> test1")
...     gr2.switch()
...     print("[gr1] test1 <- test2")
...     return 'test1 done'

>>> def test2():
...     print("[gr2] test1 -> test2")
...     gr1.switch()
...     print("This is never printed.")

>>> gr1 = greenlet(test1)
>>> gr2 = greenlet(test2)
>>> gr1.switch()
[gr1] main  -> test1
[gr2] test1 -> test2
[gr1] test1 <- test2
'test1 done'
>>> gr1.dead
True
>>> gr2.dead
False
```

**Reply by jpc1976** [reply]
This works as expected:

```
[gr1] main  -> test1
[gr2] test1 -> test2
[gr1] test1 <- test2
test1 done
True
False
```

So the problem must occur when a postgres error is raised.

**Reply by zzzeek** [reply]
try this program:


```py
from greenlet import greenlet
from greenlet import getcurrent
import asyncio
import sys

def non_awaitable(arg):
    print(f"non awaitable called with {arg}")
    return current.switch(awaitable(arg))

async def awaitable(arg):
    print(f"awaitable called with {arg}")
    if arg == "raise":
        raise Exception("oops")

    return "return value"


async def main():

    for arg in "some arg", "raise":
        gr1 = greenlet(non_awaitable)

        result = gr1.switch(arg)
        while not gr1.dead:
            try:
                value = await result
            except BaseException:
                result = gr1.throw(*sys.exc_info())
            else:
                result = gr1.switch(value)

current = getcurrent()

asyncio.run(main())
```

it should raise "oops" but not segfault.

**Reply by jpc1976** [reply]
This one did crash:

```
non awaitable called with some arg
awaitable called with some arg
non awaitable called with raise
awaitable called with raise
Segmentation fault: 11
```

Hopefully this is helpful.

**Reply by jpc1976** [reply]
Based on what I found above in https://github.com/sqlalchemy/sqlalchemy/discussions/10553#discussioncomment-7434797 with the help of @CaselIT , this appears to be related to greenlet v3 (and independent of SQLAlchemy), perhaps I should look up the issue with them.

**Reply by CaselIT** [reply]
@jpc1976 If you can report the issue to greenlet it would be great. Feel free to reference mike script and this discussion (maybe pointing to the valgrindt trace you shared). (Or if you prefer me or mike can create it)

**Reply by zzzeek** [reply]
> This one did crash:

perfect!  take that exact script along with everything about your platform as well as all valgrind / segfault metrics over to https://github.com/python-greenlet/greenlet/issues .

in fact this might be the same issue?  https://github.com/python-greenlet/greenlet/issues/382

**Reply by jpc1976** [reply]
Thanks a lot for your help guys! It does look related to https://github.com/python-greenlet/greenlet/issues/382, I see that @zzzeek has already put the link in there, but I can copy the info if you think it may be useful.

**Reply by CaselIT** [reply]
I think they would be useful, so feel free to post them!

**Reply by jpc1976** [reply]
Will do later today, thanks!
